### ISSUE TITLE
MonadError's `handleWrappedError` definition is problematic

### ISSUE THREAD

**Issue by kubukoz** [softwaremill/tapir#658]
Hi! We're migrating from 0.14 to 0.16, and previously we used to be able to change the effect of a `ServerEndpoint` quite easily. That was possible because there was no `MonadError` in the logic part:

```scala
def mapK[F[_], G[_]](endpoint: ServerEndpoint[I, E, O, S, F])(fk: F ~> G): ServerEndpoint[I, E, O, S, G] =
  endpoint.copy(logic = i => fk(endpoint.logic(i)))
```

However, now the logic function takes a MonadError - I assume it's because of the partial route implementation support built into Tapir. So our solution to the problem was to change the signature of the above method to take `F ~> G` and now also `G ~> F`, which would allow us to "wrap" the existing MonadError with one that changes the effects where necessary.

Here's what that would look like:

```scala
def imapK[F[_], G[_]](af: MonadError[F])(fk: F ~> G)(gk: G ~> F): MonadError[G] =
  new MonadError[G] {
    override def unit[T](t: T): G[T] = fk(af.unit(t))

    override def map[T, T2](fa: G[T])(f: T => T2): G[T2] = fk(af.map(gk(fa))(f))

    override def flatMap[T, T2](fa: G[T])(f: T => G[T2]): G[T2] = fk(af.flatMap(gk(fa))(f.andThen(gk(_))))

    override def error[T](t: Throwable): G[T] = fk(af.error(t))

    override protected def handleWrappedError[T](rt: G[T])(h: PartialFunction[Throwable, G[T]]): G[T] =
      fk(af.handleWrappedError(gk(rt)) {
        case t if h.isDefinedAt(t) => gk(h(t))
      })
  }
```

The problem with that is - it doesn't compile. `handleWrappedError` is `protected`, and (to my great surprise) we can't call it in a subclass of `MonadError` (whaat?). That's consistent with the answers on https://github.com/scala/bug/issues/9890 and the SLS: https://www.scala-lang.org/files/archive/spec/2.11/05-classes-and-objects.html#protected

> A protected identifier 𝑥 may be used as a member name in a selection 𝑟.𝑥
> only if one of the following applies:
> - The access is within the template defining the member, or, if a qualification 𝐶 is given, inside the package 𝐶, or the class 𝐶, or its companion module, or
> - 𝑟 is one of the reserved words this and super, or
> - *𝑟's type conforms to a type-instance of the class which contains the access.*

So we could only access `handleWrappedError` if it was both:

- A) marked as `protected[pkg]`, and 
- B) our code was also in `pkg`.

Alternatively, we could check in our call whether `af` is an instance of the same class as ours, and if it is, the last rule would apply and we could call it.

Can you please elaborate on why the `logic` function takes a MonadError instead of it being provided at the point of conversion to a route? Is it inherent to the problem, or only enforced by the particular encoding of partial routes that you chose?

In case the `logic` signature needs to remain as it is, could the access modifier (`protected`) be weakened to allow calling it by other implementations of `MonadError`? Or maybe it could be removed altogether in favor of making `handleError` abstract?

**Comment by adamw** [comment]
The `MonadError` is needed for two things: one, as you mentioned, to allow composing partial logic fragments. Second, to allow specifying the logic with error recovery (e.g. `serverLogicRecoverErrors`). This was previously part of the interpreters, but is now generalised to work with any interpreter (without duplicating the code in every one).

Wouldn't the following work for changing the effect type?

```
def mapK[F[_], G[_]](endpoint: ServerEndpoint[I, E, O, S, F])(fk: F ~> G, mf: MonadError[F]): ServerEndpoint[I, E, O, S, G] =
  endpoint.copy(logic = _ => i => fk(endpoint.logic(mf)(i)))
```

or you don't have access to the `MonadError` instance for `F` during the mapping?

The rationale of making `handleWrappedError` protected is that `handleError` is the method that should be normally used, as it takes the by-name parameter and catches exceptions that might occur when first evaluating the effect. 

I don't think we can make `handleError` - we would then need to change `handleWrappedError` to take a by-name parameter and duplicate the logic in every implementation.

So the only viable solutions seem to be either the alternative `mapK` definition, or making `handleWrappedError` public (and maybe using a better name?)

**Comment by ghostbuster91** [comment]
I am in favor of making `handleWrappedError` public, putting `serverLogicRecoverError` and the rest of the context aside, it should be possible to create an instance of `MonadError` for` G` having an instance of it for `F` when there is a NT from `F` to `G`.

**Comment by kubukoz** [comment]
> Wouldn't the following work for changing the effect type?
> 
> ```
> def mapK[F[_], G[_]](endpoint: ServerEndpoint[I, E, O, S, F])(fk: F ~> G, mf: MonadError[F]): ServerEndpoint[I, E, O, S, G] =
>   endpoint.copy(logic = _ => i => fk(endpoint.logic(mf)(i)))
> ```

It probably would - I think I can get the `MonadError` instance or build it for the concrete effect. So it's okay to ignore the one I get as input? I was worried it might change the behavior in some way I didn't expect.

>So the only viable solutions seem to be either the alternative `mapK` definition, or making `handleWrappedError` public (and maybe using a better name?)

Maybe a `mapK` could be built in? We'd need to define a `~>` type to avoid the dependency on cats, though. Still, the simplest solution for me seems to be the one with `logic = _ => i => ...`, so I'd just go for that.

**Comment by adamw** [comment]
> It probably would - I think I can get the MonadError instance or build it for the concrete effect. So it's okay to ignore the one I get as input? I was worried it might change the behavior in some way I didn't expect.

Yes, you can safely ignore it. No linear types here, the parameter can be discarded :) 

> Maybe a mapK could be built in? We'd need to define a ~> type to avoid the dependency on cats, though. Still, the simplest solution for me seems to be the one with logic = _ => i => ..., so I'd just go for that.

Maybe we could add `mapK` as an extension method provided by the cats-integration module.

Also, on second thought, maybe as you wrote we should leave `handleError` public & abstract, and provide the additional logic in sth like `handleErrorSafe`

**Comment by matwojcik** [comment]
The definition of `def mapK[F[_], G[_]](endpoint: ServerEndpoint[I, E, O, S, F])(fk: F ~> G, mf: MonadError[F]): ServerEndpoint[I, E, O, S, G]` is not compatible with `FunctorK`. 

I believe the best that could be done there is implementing `InvariantK`, because `MonadError` takes `F[_]` as method parameters, so we need `G ~> F` transformation anyway. And for that we would need protected methods to be gone from `MonadError` so we can `imapK` over them ;)

> Also, on second thought, maybe as you wrote we should leave handleError public & abstract, and provide the additional logic in sth like handleErrorSafe

Do you mean just change name `handleWrappedError` to `handleError` and `handleError` to `handleErrorSafe`, and using `handleErrorSafe` in `MonadError.recoverErrors`?
```
  def handleError[T](rt: F[T])(h: PartialFunction[Throwable, F[T]]): F[T]
  def handleErrorSafe[T](rt: => F[T])(h: PartialFunction[Throwable, F[T]]): F[T] = {
    Try(rt) match {
      case Success(v)                     => handleError(v)(h)
      case Failure(e) if h.isDefinedAt(e) => h(e)
      case Failure(e)                     => error(e)
    }
  }
```

If so I can contribute, and later on even move `InvariantK` (or rather just `imapK` functions because cats module does not depend on cats tagless) definitions for `MonadError` and `Endpoint`, that @kubukoz  mentioned, to tapir cats module.

**Comment by ghostbuster91** [comment]
As `handleWrappedError` is only called from `handleError` and the only one place where we call `handleError` is from `recoverErrors` method why just don't inline `handleError` method there? So it would look like follows:
```
def recoverErrors[I, E, O, F[_]](
      f: I => F[O]
  )(implicit eClassTag: ClassTag[E], eIsThrowable: E <:< Throwable): MonadError[F] => I => F[Either[E, O]] = { implicit monad => i =>
    import sttp.tapir.monad.syntax._
    Try(f(i).map(Right(_): Either[E, O])) match {
      case Success(value) =>
        monad.handleError(value) {
          case e if eClassTag.runtimeClass.isInstance(e) => wrapException(e)
        }
      case Failure(exception) if eClassTag.runtimeClass.isInstance(exception) => wrapException(exception)
      case Failure(exception)                                                 => monad.error(exception)
    }
  }

  private def wrapException[F[_], O, E, I](exception: Throwable)(implicit me: MonadError[F]) = {
    me.unit(Left(exception.asInstanceOf[E]): Either[E, O])
  }
```

Then we will be able to remove `handleError` from `MonadError` and rename `handleWrappedError` to `handleError`:
```
trait MonadError[F[_]] {
  def unit[T](t: T): F[T]
  def map[T, T2](fa: F[T])(f: T => T2): F[T2]
  def flatMap[T, T2](fa: F[T])(f: T => F[T2]): F[T2]

  def error[T](t: Throwable): F[T]
  def handleError[T](rt: => F[T])(h: PartialFunction[Throwable, F[T]]): F[T]
}
```
Which will allow us finally to `imapK` over it :)

:+1: for defining `imapK` functions for `MonadError` and `Endpoint`

**Comment by adamw** [comment]
@matwojcik yes, that's what I had in mind :)

@ghostbuster91 good point :)

**Comment by matwojcik** [comment]
I've pushed suggestions in https://github.com/softwaremill/tapir/pull/663 together with cats syntax

**Comment by adamw** [comment]
Released in 0.16.2
